lib/sms/sms_at.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <modem/sms.h>
+#include <errno.h>
+#include <modem/at_cmd.h>
+#include <modem/at_cmd_parser.h>
+#include <modem/at_params.h>
+#include <modem/at_notif.h>
+
+#include "string_conversion.h"
+#include "parser.h"
+#include "sms_submit.h"
+#include "sms_deliver.h"
+#include "sms_internal.h"
+
+LOG_MODULE_DECLARE(sms, CONFIG_SMS_LOG_LEVEL);
+
+/** @brief Number of parameters in CMT AT notification. */
+#define AT_CMT_PARAMS_COUNT 4
+/* SMS PDU is a bit below 180 bytes which means 360 two character long
+ * hexadecimal numbers. In addition CMT response has alpha and length so
+ * reserving safely a bit over maximum.
+ */
+#define AT_CMT_PDU_MAX_LEN 512
+/** @brief PDU index in CMT AT notification. */
+#define AT_CMT_PDU_INDEX 3
+/** @brief Number of parameters in CDS AT notification. */
+#define AT_CDS_PARAMS_COUNT 3
+/** @brief PDU index in CDS AT notification. */
+#define AT_CDS_PDU_INDEX 2
+
+/** @brief Start of AT notification for incoming SMS. */
+#define AT_SMS_DELIVER "+CMT:"
+/** @brief Length of AT notification for incoming SMS. */
+#define AT_SMS_DELIVER_LEN (sizeof(AT_SMS_DELIVER) - 1)
+
+/** @brief Start of AT notification for incoming SMS status report. */
+#define AT_SMS_STATUS_REPORT "+CDS:"
+/** @brief Length of AT notification for incoming SMS status report. */
+#define AT_SMS_STATUS_REPORT_LEN (sizeof(AT_SMS_STATUS_REPORT) - 1)
+
+
+/**
+ * @brief Parses AT notification and finds PDU buffer.
+ *
+ * @param[in] buf AT notification buffer.
+ * @param[out] pdu Output buffer where PDU is copied.
+ * @param[in] pdu_len Length of the output buffer.
+ * @param[in] at_params_count Maximum number of AT parameters in AT notification.
+ * @param[in] pdu_index Index where PDU is found from AT notification parameter list.
+ * @param[in] temp_resp_list Response list used by AT parser library. This is readily initialized
+ *                 by caller and is passed here to avoid using another instance of the list.
+ * @return Zero on success and negative value in error cases.
+ */
+static int sms_notif_at_parse(const char *const buf, char *pdu, size_t pdu_len,
+	int at_params_count, int pdu_index, struct at_param_list *temp_resp_list)
+{
+	int err = at_parser_max_params_from_str(buf, NULL, temp_resp_list, at_params_count);
+
+	if (err != 0) {
+		LOG_ERR("Unable to parse AT notification, err=%d", err);
+		return err;
+	}
+
+	(void)at_params_string_get(temp_resp_list, pdu_index, pdu, &pdu_len);
```

Comment:
You are right. Fixed.

Reply:
Hmmm, now when trying to do a test for the added if-statement, it's likely not possible because at_parser library handles the fact that the parameter is string type for SMS PDU by having special treatment for SMS at notifications.

Reply:
Finally found a way to trigger this new check with external input.
Good catch!